streams/src/test/java/org/apache/kafka/streams/integration/QueryableStateIntegrationTest.java

```diff
@@ -115,8 +162,216 @@ public void shutdown() throws IOException {
         IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
     }
 
+
+    /**
+     * Creates a typical word count topology
+     * @param inputTopic
+     * @param outputTopic
+     * @param streamsConfiguration config
+     * @return
+     */
+    private KafkaStreams createCountStream(String inputTopic, String outputTopic, Properties streamsConfiguration) {
+        KStreamBuilder builder = new KStreamBuilder();
+        final Serde<String> stringSerde = Serdes.String();
+        final KStream<String, String> textLines = builder.stream(stringSerde, stringSerde, inputTopic);
+
+        final KGroupedStream<String, String> groupedByWord = textLines
+            .flatMapValues(new ValueMapper<String, Iterable<String>>() {
+                @Override
+                public Iterable<String> apply(String value) {
+                    return Arrays.asList(value.toLowerCase(Locale.getDefault()).split("\\W+"));
+                }
+            })
+            .groupBy(MockKeyValueMapper.<String, String>SelectValueMapper());
+
+        // Create a State Store for the all time word count
+        groupedByWord.count("word-count-store-" + inputTopic).to(Serdes.String(), Serdes.Long(), outputTopic);
+
+        // Create a Windowed State Store that contains the word count for every 1 minute
+        groupedByWord.count(TimeWindows.of(WINDOW_SIZE), "windowed-word-count-store-" + inputTopic);
+
+        return new KafkaStreams(builder, streamsConfiguration);
+    }
+
+    private class StreamRunnable implements Runnable {
+        private final KafkaStreams myStream;
+        private boolean closed = false;
+        StreamRunnable(String inputTopic, String outputTopic, int queryPort) {
+            Properties props = (Properties) streamsConfiguration.clone();
+            props.put(StreamsConfig.APPLICATION_SERVER_CONFIG, "localhost:" + queryPort);
+            this.myStream = createCountStream(inputTopic, outputTopic, props);
+        }
+
+        @Override
+        public void run() {
+            this.myStream.start();
+
+        }
+
+        public void close() {
+            if (!closed) {
+                this.myStream.close();
+                closed = true;
+            }
+        }
+
+        public boolean isClosed() {
+            return closed;
+        }
+
+        public final KafkaStreams getStream() {
+            return myStream;
+        }
+    }
+
+    private void verifyAllKVKeys(final StreamRunnable[] streamRunnables, final KafkaStreams streams,
+                                 final Set<String> keys, final Set<String> stateStores) throws Exception {
+        for (final String storeName : stateStores) {
+            for (final String key : keys) {
+                TestUtils.waitForCondition(new TestCondition() {
```

Comment:
The verification latency is dependent on the ordering of keys, maybe we can optimize this a bit by keeping a bit array indicating if we have seen this key or not; and call `watForCondition` outside of the loops checking the bit array, while inside the `conditionMet` while loop the keys whose bits are not set to check if `store != null && store.get(key) != null`.

Ditto below.